tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,290 @@
+            }
+            finally
+            {
+                SemaphoreSlim.Release();
+            }
+        }
+
+        private GetRcmRequest BuildRequest()
+        {
+            var appliedConfigurations = _products.Values.SelectMany(pair => pair.AppliedConfigurations.Values).ToList();
+            var cachedTargetFiles = appliedConfigurations.Select(cache => new RcmCachedTargetFile(cache.Path.Path, cache.Length, cache.Hashes.Select(kp => new RcmCachedTargetFileHash(kp.Key, kp.Value)).ToList())).ToList();
+
+            return new GetRcmRequest(BuildRequestClient(), cachedTargetFiles);
+
+            RcmClient BuildRequestClient()
+            {
+                return new RcmClient(_id, _products.Keys.ToList(), BuildRcmClientTracer(), BuildRequestClientState());
+            }
+
+            RcmClientTracer BuildRcmClientTracer()
+            {
+                return new RcmClientTracer(_runtimeId, _tracerVersion, _service, _env, _appVersion);
+            }
+
+            RcmClientState BuildRequestClientState()
+            {
+                var configStates = appliedConfigurations.Select(cache => new RcmConfigState(cache.Path.Id, cache.Version, cache.Path.Product)).ToList();
+                return new RcmClientState(_rootVersion, _targetsVersion, configStates, _lastPollError != null, _lastPollError);
+            }
+        }
+
+        private void ProcessResponse(GetRcmResponse response)
+        {
+            var actualConfigPath =
+                response.ClientConfigs
+                        .Select(RemoteConfigurationPath.FromPath)
+                        .ToDictionary(path => path.Path);
+
+            var changedConfigurationsByProduct = GetChangedConfigurations()
+               .GroupBy(config => config.Path.Product);
+
+            foreach (var productGroup in changedConfigurationsByProduct)
+            {
+                var product = _products[productGroup.Key];
+
+                try
+                {
+                    var configurations = productGroup.ToList();
+
+                    product.AssignConfigs(configurations);
+                    CacheAppliedConfigurations(product, configurations);
+                }
+                catch (Exception e)
+                {
+                    Log.Warning($"Failed to apply remote configurations {e.Message}");
+                }
+            }
+
+            UnapplyRemovedConfigurations();
+
+            IEnumerable<RemoteConfiguration> GetChangedConfigurations()
+            {
+                var signed = response.Targets.Signed.Targets;
+                var targetFiles = (response.TargetFiles ?? Enumerable.Empty<RcmFile>()).ToDictionary(f => f.Path, f => f);
+
+                foreach (var kp in actualConfigPath)
+                {
+                    if (!signed.TryGetValue(kp.Key, out var signedTarget))
+                    {
+                        throw new RemoteConfigurationException($"Missing config {kp.Key} in targets");
```

Comment:
nit - We could use a ThrowHelper instead

Reply:
done at 617a7dd73da9a861b4f39fefe1b8478535994d7b